TrustKit fix

rclee opened the pull request:
Changes to TrustKit to fix [nullability warnings](https://github.com/mapbox/mapbox-events-ios/issues/66) in Xcode 10/iOS 12 and fix [a data race](https://github.com/mapbox/mapbox-events-ios/issues/71) with TrustKit's logging. The data race issue is only an issue in non-release builds, in release builds the logger is disabled.

rclee committed:
Add fix for Trustkit

Pulls in changes from TrustKit PR. Also updates project to match Xcode 10 settings

ref: https://github.com/mapbox/mapbox-events-ios/issues/66

rclee committed:
Call URLSession:didReceiveChallenge: completion block on the main queue

This is to fix an error on TrustKit logger tests

friedbunny reviewed (dismissed):
Thanks! If possible, it’d be nice to get this into `gl-native` before tomorrow’s `ios-v4.5.0-beta.1`.

rclee committed:
Ensure handleChallenge: is called on a background queue

rclee committed:
Expose and call didReceiveChallenge: directly

WIP

rclee committed:
Add WIP

rclee committed:
Fix didRecieveChallenge: tests

Adds cedar tests for didRecieveChallenge: to confirm we are returning completion on the main thread when appropriate.

rclee commented:
Pushed up the changes we worked on @julianrex and added an additional test for when TrustKit will evaluate a challenge. Would be great to get  👀 on this PR again. Hopefully we can put this horse out to pasture... 🐴

rclee committed:
Merge branch 'master' into rclee-trustkit-fix

julianrex commented on MapboxMobileEventsTests/MMEAPIClientTests.mm:
```diff
@@ -43,6 +47,136 @@ - (void)setupUserAgent;
+            });
+            
+            it(@"should be on background queue", ^{
+                isMainThread should be_falsy;
```
i.e. I think this `should be_truthy`

julianrex commented on MapboxMobileEvents/MMENSURLSessionWrapper.m:
```diff
@@ -46,12 +46,36 @@ - (void)processRequest:(NSURLRequest *)request completionHandler:(void (^)(NSDat
+    __block void (^completion)(NSURLSessionAuthChallengeDisposition, NSURLCredential * _Nullable) = nil;
+    
+    completion = ^(NSURLSessionAuthChallengeDisposition disposition, NSURLCredential * _Nullable credential){
+        dispatch_async(dispatch_get_main_queue(), ^{
```
For comments below - the completion block is dispatched to the main queue (on the main thread), which is what we want.

rclee committed:
Remove APIServiceURLProtocol

Don't need it for these tests

rclee committed:
Correct test and clean-up

julianrex reviewed (dismissed):
Looks good from the code side. You may want someone more familiar with Cedar to check on the tests.

rclee commented:
Thanks so much @julianrex 🙇 -- @akitchen ? Mind giving it a once over? 💚

akitchen commented on MapboxMobileEventsTests/MMEAPIClientTests.mm:
```diff
@@ -43,6 +46,136 @@ - (void)setupUserAgent;
+                
+                while (!completionBlockCalled) {
+                    NSDate *futureDate = [NSDate dateWithTimeIntervalSinceNow:0.1];
+                    [[NSRunLoop currentRunLoop] runUntilDate:futureDate];
```
if this test were to break it would spin the while loop forever, which may or may not be a concern. have you considered failing it after a timeout?

akitchen commented:
Looks good to me, although you might consider either a synchronous dispatch or using a dispatch_semaphore instead of the while loops in the tests.

rclee committed:
Remove potential infinite looping

rclee commented:
I've added dispatch_semaphore and this seems to be operating as expected. Let me know how that looks 👀

rclee committed:
Merge branch 'master' into rclee-trustkit-fix

rclee committed:
No infinite loops

rclee committed:
Merge branch 'master' into rclee-trustkit-fix

rclee committed:
Merge remote-tracking branch 'origin/rclee-trustkit-fix' into rclee-trustkit-fix

rclee committed:
Use variable for nanoseconds

alfwatt commented on MapboxMobileEventsTests/MMEAPIClientTests.mm:
```diff
@@ -43,6 +47,145 @@ - (void)setupUserAgent;
+                isMainThread should be_truthy;
+            });
+            
+            it(@"should receive challenge", ^{
```
Good point: https://github.com/mapbox/mobile-telemetry/issues/291

rclee replied:
Fair point, these tests stopped being valuable when we changed directions and I failed to notice that. I'll make a new PR to fix this.
